File: Pandacap.HighLevel/LemmyClient.cs
using Microsoft.FSharp.Collections;
using Pandacap.LowLevel;
using System.Runtime.CompilerServices;

namespace Pandacap.HighLevel
{
    public class LemmyClient(
        IHttpClientFactory httpClientFactory)
    {
        public async Task<Lemmy.Community> GetCommunityAsync(
            string host,
            string name,
            CancellationToken cancellationToken = default)
        {
            using var client = httpClientFactory.CreateClient();
            client.DefaultRequestHeaders.UserAgent.ParseAdd(UserAgentInformation.UserAgent);
            var resp = await Lemmy.GetCommunityAsync(client, host, name, cancellationToken);
            return resp.community_view.community;
        }

        public async Task<(Lemmy.PostView, Lemmy.Community)> GetPostAsync(
            string host,
            int id,
            CancellationToken cancellationToken = default)
        {
            using var client = httpClientFactory.CreateClient();
            client.DefaultRequestHeaders.UserAgent.ParseAdd(UserAgentInformation.UserAgent);
            var resp = await Lemmy.GetPostAsync(client, host, id, cancellationToken);
            return (resp.post_view, resp.community_view.community);
        }

        public async Task<FSharpList<Lemmy.PostView>> GetPostsAsync(
            string host,
            int community_id,
            Lemmy.GetPostsSort sort,
            int page = 1,
            int limit = 10,
            CancellationToken cancellationToken = default)
        {
            using var client = httpClientFactory.CreateClient();
            client.DefaultRequestHeaders.UserAgent.ParseAdd(UserAgentInformation.UserAgent);

            var resp = await Lemmy.GetPostsAsync(
                client,
                host,
                [
                    Lemmy.GetPostsParameter.NewSort(sort),
                    Lemmy.GetPostsParameter.NewPage(page),
                    Lemmy.GetPostsParameter.NewLimit(limit),
                    Lemmy.GetPostsParameter.NewCommunityId(community_id)
                ],
                cancellationToken);

            return resp.posts;
        }

        public async IAsyncEnumerable<Lemmy.CommentObject> GetCommentsAsync(
            string host,
            int post_id,
            Lemmy.GetCommentsSort sort,
            [EnumeratorCancellation] CancellationToken cancellationToken = default)
        {
            using var client = httpClientFactory.CreateClient();
            client.DefaultRequestHeaders.UserAgent.ParseAdd(UserAgentInformation.UserAgent);

            int page = 1;

            while (true)
            {
                var resp = await Lemmy.GetCommentsAsync(
                    client,
                    host,
                    [
                        Lemmy.GetCommentsParameter.NewSort(sort),
                        Lemmy.GetCommentsParameter.NewPostId(post_id),
                        Lemmy.GetCommentsParameter.NewPage(page)
                    ],
                    cancellationToken);

                if (resp.comments.Length == 0)
                    break;

                foreach (var comment in resp.comments)
                    yield return comment;

                page++;
            }
        }
    }
}

File: Pandacap.Lemmy/LemmyClient.fs
namespace Pandacap.Lemmy

open System
open System.Net.Http
open System.Net.Http.Json
open FSharp.Control
open Pandacap.Lemmy.Models
open Pandacap.Lemmy.Interfaces

type LemmyClient(
    httpClientFactory: IHttpClientFactory
) =
    interface ILemmyClient with
        member _.GetCommentsAsync(host, post_id, sort) = asyncSeq {
            use client = httpClientFactory.CreateClient()

            let mutable page = 1
            let mutable finished = false

            while not finished do
                let! resp = Lemmy.asyncGetComments client host [
                    Sort sort
                    PostId post_id
                    Page page
                ]

                yield! resp.comments

                if List.isEmpty resp.comments then
                    finished <- true

                page <- page + 1
        }

        member _.GetCommunityAsync(host, name, cancellationToken) = task {
            use client = httpClientFactory.CreateClient()

            let! communityResponse = Async.StartAsTask(
                Lemmy.asyncGetCommunity client host name,
                cancellationToken = cancellationToken)

            return communityResponse.community_view.community
        }

        member _.GetPostAsync(host, id, cancellationToken) = task {
            use client = httpClientFactory.CreateClient()

            let! postResponse = Async.StartAsTask(
                Lemmy.asyncGetPost client host id,
                cancellationToken = cancellationToken)

            return struct (postResponse.post_view, postResponse.community_view.community)
        }

        member _.GetPostsAsync(host, community_id, sort, start_page) = asyncSeq {
            use client = httpClientFactory.CreateClient()

            let mutable page = start_page
            let mutable finished = false

            while not finished do
                let! resp = Lemmy.asyncGetPosts client host [
                    GetPostsParameter.Sort sort
                    GetPostsParameter.Page page
                    GetPostsParameter.Limit 10
                    GetPostsParameter.CommunityId community_id
                ]

                yield! resp.posts

                if List.isEmpty resp.posts then
                    finished <- true

                page <- page + 1
        }

        member _.Restructure(comments) = Lemmy.restructure comments
